### PR TITLE
Hover show documentation for XSI attributes

### LINKED ISSUE #164: Hover for xsi attributes
Add hover for xsi attributes:

 - `schemaLocation`
 - `noNamespaceSchemaLocation`
 -  `nil`?

See https://www.w3.org/TR/xmlschema-1/#d0e3067

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #164

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 1]
```diff
@@ -12,24 +12,48 @@
...
 /**
  * This class holds values that represent the XSI xsd.
  * Can be seen at https://www.w3.org/2001/XMLSchema-instance
  */
 public class XSISchemaModel {
-	
+
+	private static final String typeDoc = "Specifies the type of an element. This attribute labels an element as a particular type, even though there might not be an element declaration in the schema binding that element to the type."; 
+	private static final String nilDoc = "Indicates if an element should contain content. Valid values are true/false";
+	private static final String schemaLocationDoc =
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: line separators should be platform specific

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 2]
```diff
@@ -12,24 +12,48 @@
...
+import org.eclipse.lsp4xml.services.extensions.IHoverRequest;
 
 /**
  * This class holds values that represent the XSI xsd.
  * Can be seen at https://www.w3.org/2001/XMLSchema-instance
  */
 public class XSISchemaModel {
-	
+
+	private static final String typeDoc = "Specifies the type of an element. This attribute labels an element as a particular type, even though there might not be an element declaration in the schema binding that element to the type.";
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: use all caps for static constants

Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: static fields could be made package visible so they can be referenced from tests directly

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 3]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
 
+	@Test
+	public void testTagHoverForSchemaLocation() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
+		String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n" + //
+				"<invoice xmlns=\"http://invoice\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ " xsi:schema|Location=\"http://invoice xsd/invoice.xsd \">\r\n";
+		String doc =
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: reference actual doc constant instead

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 4]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
+				doc, null);
+
+	};
+
+	@Test
+	public void testTagHoverForXSIBadPrefix() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 5]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+		String doc = "Specifies the type of an element. This attribute labels an element as a particular type, even though there might not be an element declaration in the schema binding that element to the type."; 
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
+				doc, null);
+	};
+
+	@Test
+	public void testTagHoverForXSIType2() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 6]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+
+	@Test
+	public void testTagHoverForXSIType() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
+		String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n" + //
+				"<invoice xmlns=\"http://invoice\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ " xsi:ty|pe=\"http://invoice xsd/invoice.xsd \">\r\n";
+		String doc = "Specifies the type of an element. This attribute labels an element as a particular type, even though there might not be an element declaration in the schema binding that element to the type.";
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: reference actual doc constant instead

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 7]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
+				doc, null);
+
+	};
+
+	@Test
+	public void testTagHoverForXSIType() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 8]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+
+	@Test
+	public void testTagHoverForXSIType2() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
+		String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n" + //
+				"<invoice xmlns=\"http://invoice\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ " x|si:type=\"http://invoice xsd/invoice.xsd \">\r\n";
+		String doc = "Specifies the type of an element. This attribute labels an element as a particular type, even though there might not be an element declaration in the schema binding that element to the type.";
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: reference actual doc constant instead

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 9]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+
+	@Test
+	public void testTagHoverForXSInil() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
+		String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n" + //
+				"<invoice xmlns=\"http://invoice\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ " xsi:n|il=\"http://invoice xsd/invoice.xsd \">\r\n";
+		String doc = "Indicates if an element should contain content. Valid values are true/false";
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: reference actual doc constant instead

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 10]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
+				doc, null);
+
+	};
+
+	@Test
+	public void testTagHoverForXSInil() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 11]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+
+	@Test
+	public void testTagHoverForNoNamespaceSchemaLocation() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
+		String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n" + //
+				"<invoice xmlns=\"http://invoice\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ " xsi:noNamespace|SchemaLocation=\"http://invoice xsd/invoice.xsd \">\r\n";
+		String doc =
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: reference actual doc constant instead

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 12]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
...
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
+				doc, null);
+
+	};
+
+	@Test
+	public void testTagHoverForNoNamespaceSchemaLocation() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java [hunk 13]
```diff
@@ -56,6 +56,120 @@ public void testTagHoverFromXSType() throws BadLocationException {
 
 	};
 
+	@Test
+	public void testTagHoverForSchemaLocation() throws BadLocationException {
+		// web.xml servlet, servlet-name declares their xs:annotation not in the element
+		// declaration but in type (servletType),
+		// this test check that
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java]: checks

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 14]
```diff
@@ -102,4 +113,35 @@ private static void createCompletionItem(String text, boolean isSnippetsSupporte
 	private static boolean attributeAlreadyExists(DOMElement root, String actualPrefix, String suffix) {
 		return root.getAttributeNode(actualPrefix + ":" + suffix) != null; 
 	}
+
+	public static Hover computeHoverResponse(DOMAttr attribute, IHoverRequest request) {
+		if(!request.getXMLDocument().getDocumentElement().equals(request.getXMLDocument().findNodeAt(attribute.getStart()))) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: is everything null-safe here?

**Comment by NikolasKomonen** [comment]
@fbricon updated

**Comment by NikolasKomonen** [comment]
@fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java [hunk 15]
```diff
@@ -54,6 +55,10 @@ public Hover onAttributeName(IHoverRequest hoverRequest) throws Exception {
 		try {
 			ContentModelManager contentModelManager = hoverRequest.getComponent(ContentModelManager.class);
 			DOMAttr attribute = (DOMAttr) hoverRequest.getNode();
+			Hover temp = XSISchemaModel.computeHoverResponse(attribute, hoverRequest);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java]: need sa comment explaining why it's not handled like other hovers through contentModelManager. findCMElement

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java [hunk 16]
```diff
@@ -54,6 +55,10 @@ public Hover onAttributeName(IHoverRequest hoverRequest) throws Exception {
 		try {
 			ContentModelManager contentModelManager = hoverRequest.getComponent(ContentModelManager.class);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java]: only used if not an xsi attribute. Hence need to be moved after L61

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 17]
```diff
@@ -43,55 +78,42 @@ public static void computeCompletionResponses(ICompletionRequest request,
...
 	private static void createCompletionItem(String text, boolean isSnippetsSupported, 
-			String defaultAndSnippet, Range editRange, ICompletionResponse response, String documentation) {
+			String defaultAndSnippet, Range editRange, ICompletionResponse response, String documentation, boolean generateValue) {
 		CompletionItem item = new CompletionItem();
 		item.setLabel(text);
 		item.setKind(CompletionItemKind.Value);
 		item.setFilterText(text);
 		item.setInsertTextFormat(isSnippetsSupported ? InsertTextFormat.Snippet : InsertTextFormat.PlainText);
-		item.setTextEdit(new TextEdit(editRange, text + "=\"" + defaultAndSnippet + "\""));
+		text = generateValue ? text + "=\"" + defaultAndSnippet + "\"" : text + "$0";
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: $0 placeholder can only be used if snippets are supported

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 18]
```diff
@@ -102,4 +124,45 @@ private static void createCompletionItem(String text, boolean isSnippetsSupporte
...
+				doc = TYPE_DOC;
+			}
+			else {
+				return null;
+			}
+		}
+		
+
+		MarkupContent content = new MarkupContent();
+		content.setKind(MarkupKind.MARKDOWN);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: what bothers me here is the inconsistency between this hover being markdown, others being plaintext. It's probably tolerable but we should probably try to return markdown for other documentation. That's the subject of another ticket though.

Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: Opened #245

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java (hunk 17 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: Please note I will push a new class AttributeCompletionItem will simplify the creation of completion item for attribute (takes care of snippet support, generate choice snippet if attributes has values (like "true", "false")

**Comment by NikolasKomonen** [comment]
@fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java [hunk 19]
```diff
@@ -12,74 +12,98 @@
 
 import java.util.Collection;
 
+import org.apache.xml.serialize.LineSeparator;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XSISchemaModel.java]: organize imports here

**Comment by NikolasKomonen** [comment]
@fbricon updated

**Pull request merged**